stdlib/string.ml

```diff
@@ -123,6 +123,51 @@ let escaped s =
   in
   escape_if_needed s (length s) 0
 
+let escape_ascii = escaped
+
+let escape s =
+  (* Escape only C0 control characters (bytes <= 0x1F), DEL(0x7F), '\\'
+     and '"' *)
+   let n = ref 0 in
+  for i = 0 to length s - 1 do
+    n := !n +
+      (match unsafe_get s i with
+       | '\"' | '\\' | '\n' | '\t' | '\r' | '\b' -> 2
+       | '\x00' .. '\x1F'
+       | '\x7F' -> 4
+       | _ -> 1)
+  done;
+  if !n = length s then s else begin
+    let s' = B.create !n in
+    n := 0;
+    for i = 0 to length s - 1 do
+      begin match unsafe_get s i with
+      | ('\"' | '\\') as c ->
+          B.unsafe_set s' !n '\\'; incr n; B.unsafe_set s' !n c
+      | '\n' ->
+          B.unsafe_set s' !n '\\'; incr n; B.unsafe_set s' !n 'n'
+      | '\t' ->
+          B.unsafe_set s' !n '\\'; incr n; B.unsafe_set s' !n 't'
+      | '\r' ->
+          B.unsafe_set s' !n '\\'; incr n; B.unsafe_set s' !n 'r'
+      | '\b' ->
+          B.unsafe_set s' !n '\\'; incr n; B.unsafe_set s' !n 'b'
+      | '\x00' .. '\x1F' | '\x7F' as c ->
+          let a = Char.code c in
+          B.unsafe_set s' !n '\\';
+          incr n;
+          B.unsafe_set s' !n (Char.chr (48 + a / 100));
+          incr n;
+          B.unsafe_set s' !n (Char.chr (48 + (a / 10) mod 10));
+          incr n;
+          B.unsafe_set s' !n (Char.chr (48 + a mod 10));
+      | c -> B.unsafe_set s' !n c
+      end;
+      incr n
+    done;
+    B.to_string s'
+  end
```

Comment:
Why not share an implementation with `String.escaped` and `Bytes.escaped`?  It's the same loops, all you need is a Boolean to say whether `'\x80'...'\xFF'` needs escaping or not.

Reply:
I agree I guess. That function was duplicated in the toplevel implementation, I simply moved it to the standard library.

Do I get your plan right?

* Add a new function like `val _escaped : bool → string → string` with a boolean flag to indicate whether to escape unicode or not, but don't expose it in the module interface.
* Add `escaped_ascii` and `escaped_utf8` wrappers for it?
* Alias `escaped` to `escape_ascii`.

Reply:
Yes, except that you also need to update the `Bytes` module...  So, the new `bool -> string -> string` function probably goes in `Bytes`, and is exported but not documented, so that it can be used from `String` as well.

Reply:
What do you think of adding an optional argument `?(escape_unicode=true)` to `String.escaped` and making the new `escaped_ascii` and `escaped_utf8` functions aliases for it with different value of that argument?

Reply:
I like optional arguments, generally speaking.  But here, with an optional argument on `escaped`, it would be strange to provide `escaped_ascii` and `escaped_utf8` in addition to `escaped`. Also, putting a deprecation warning on `escaped` (regardless of optional argument) would make little sense.

Reply:
I imagine that we could have an optional argument with the parameter `escape_unicode` set as true by default at first, and then switch to `false` instead of deprecating the `escaped` function few version later.

Reply:
If I were to decide, I'd go for an optional argument.

Also, tangentially related: there's code duplication with `Char.escaped`. Is there a reason for it, or we can make `Char.escaped` a building block for this?